File: ProblemSolve/56.cpp
/*******************
 * Merge Intervals *
 *******************/

class Solution {
public:
    vector<vector<int>> merge(vector<vector<int>>& intervals) {
        vector<vector<int>> ans;        
        //시작점을 기준으로 정렬
        sort(intervals.begin(), intervals.end());
        
        int len = intervals.size();
        //제일 처음이 되는 값들 push_back
        ans.push_back({intervals[0][0], intervals[0][1]});
        
        for(int i=1; i<len; i++){
            //끝나는 시간이 시작 시간보다 작으면
            //정상적으로 다음 값이 들어와야 한다
            if(ans.back()[1] < intervals[i][0]){
                ans.push_back({intervals[i][0], intervals[i][1]});
            }
            //끝나는 시간이 다음 시작시간과 겹치면
            //현재 끝나는 시간과 들어올 끝나는 시간을 비교해 큰 값을 넣어준다
            else{
                ans.back()[1] = max(ans.back()[1], intervals[i][1]);
            }
        }
        
        return ans;
    }
};

File: ProblemSolve/56.cs
/*******************
 * Merge Intervals *
 *******************/

public class Solution {
    public void Swap(ref int[] a, ref int[] b){
        int[] temp = a;
        a = b;
        b = temp;
    }
    
    public void QuickSort(ref int[][] intervals, int start, int end){
        if(start >= end){
            return;
        }
        
        ref int[] pivot = ref intervals[end];
        int left = start;
        int right;
        
        for(right = start; right<end; ++right){
            if(intervals[right][0] < pivot[0]){
                Swap(ref intervals[right], ref intervals[left]);
                ++left;
            }
        }
        Swap(ref intervals[left], ref pivot);

	    QuickSort(ref intervals, left + 1, end);
	    QuickSort(ref intervals, start, left - 1);

	    return;
    }
    
    public int[][] Merge(int[][] intervals) {
        List<int[]> ans = new List<int[]>();
        int len = intervals.Length;
        
        QuickSort(ref intervals, 0, len-1);     //intervals = intervals.OrderBy(x => x[0]).ToArray();
        
        ans.Add(new int[2]{intervals[0][0], intervals[0][1]});
        
        for(int i=1; i<len; ++i){
            //새로운 범위 정답에 추가
            if(ans[ans.Count-1][1] < intervals[i][0]){
                ans.Add(new int[2]{intervals[i][0], intervals[i][1]});
            }
            //범위가 겹치면 합치기
            else{
                ans[ans.Count-1][1] = Math.Max(intervals[i][1], ans[ans.Count-1][1]);
            }
        }
                                              
        return ans.ToArray();
    }
}